### ISSUE TITLE
Bug: [@typescript-eslint/no-invalid-void-type] void is only valid as a return type or generic type argument

### ISSUE THREAD

**Issue by splincode** [typescript-eslint/typescript-eslint#6547]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.2&fileType=.ts&code=FAQwRgzgLgTiDGUAE8A2IISQQUrBUAKgK4CWAIqSKgPYDmAygKYwBup8TAPANIB8SAN7AkSGgAcmAOy4BxJAF4krGqQAmfABQjRKGlKjSoALiQ8kTAB6GparCvVIA-EnmmeOgJTuL16XeVVNWdXJB9hXSQYJihiGCk9AyMAbh0AX2AM4CA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2HImjQO0SABp5CrAuyQB8DgHcAkkwDiiJkspkAKiUQBBaAHNYAWyv5UGfHEQtbABfIIBdeVDgoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
abstract class AbstractTuiDialogService<K> {
  open<G = void>(
    content: K extends void ? G : K
  ): K extends void ? G : K {
    return content;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    '@typescript-eslint/no-invalid-void-type': [
                    'error',
                    {
                        allowInGenericTypeArguments: true,
                    },
                ]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

error

### Actual Result

No error for generic

**Comment by biplavmz** [comment]
Hii,
you can write void like this ()=> void it will not throw any error and working fine 

abstract class AbstractTuiDialogService<K> {
  open<G = void>(
    content: K extends () => void ? G : K
  ): K extends () => void ? G : K {
    return content;
  }
}

**Comment by splincode** [comment]
@bradzacher But I need `K extends void` instead of `K extends () => void`

**Comment by armano2** [comment]
```ts
const x = new AbstractTuiDialogService<string>()

// (method) AbstractTuiDialogService<string>.open<number>(content: string): string
x.open<number>('')

// (method) AbstractTuiDialogService<string>.open<number>(content: string): string
x.open<string>('')

// (method) AbstractTuiDialogService<string>.open<void>(content: string): string
x.open<void>('')
```

```
// (method) AbstractTuiDialogService<void>.open<number>(content: number): number
x.open<number>('')

// (method) AbstractTuiDialogService<void>.open<string>(content: string): string
x.open<string>('')

// (method) AbstractTuiDialogService<void>.open<void>(content: void): void
x.open<void>('')
```

**Comment by splincode** [comment]
@armano2 https://github.com/Tinkoff/taiga-ui/blob/b8ad465a9c1f1746662afabe05306c8df54bdda5/projects/cdk/abstract/dialog.service.ts#L24

**Comment by felipesantoz** [comment]
@splincode looking at your code, I don't really see the point of having both the variables `G` and `K` - they seem to only be used for the exact same contexts, so why not just get rid of one of them and simplify your code?

**Comment by splincode** [comment]
@felipesantoz https://github.com/Tinkoff/taiga-ui/blob/b8ad465a9c1f1746662afabe05306c8df54bdda5/projects/cdk/abstract/dialog.service.ts#L24

**Comment by felipesantoz** [comment]
@splincode Yes I've viewed that link, the generics are redundant there as well

**Comment by simonbuchan** [comment]
A minimal reproduction case:

```ts
declare function foo<R>(): void;
foo<void>();
```
https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwB4AlAPgAoBKALngDcctgBuAKEUKMecqvaA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2HImjQO0SABp5CrAt179UAfA4B3AJJMA4oiZLKZAColEAQWgBzWAFsb+VBkh8OERILWwAXzCAXXlI8KA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

Seems there were quite a few ways to use void in generic arguments missed?

**Comment by ddubrava** [comment]
It looks like these things have been intentionally blocked

![image](https://github.com/typescript-eslint/typescript-eslint/assets/22116465/035c081b-15b6-41d9-b81a-579a82969637)
![image](https://github.com/typescript-eslint/typescript-eslint/assets/22116465/9a2b8f4b-4958-4d13-a8da-cc0b98a854c1)

I can definitely work on this issue if @bradzacher clarifies what the expected result should be. Allowing 'void' in call expressions and 'extends' in type parameters doesn't seem too tricky. The real problem is, how should we handle 'extends void' in conditional types? And should we simply allow 'void' in call expressions, or should we also check for a generic type in the function's type parameters?

**Comment by MatthiasKunnen** [comment]
I have the following function where I run into this issue

```typescript
function foo<T>(): (input: T) => void {
    return (value) => {
        // Implementation
    };
}

foo<void>();
//  ^^^^ ESLint: void is only valid as a return type or generic type argument.
``` 

This seems on purpose according to https://github.com/typescript-eslint/typescript-eslint/blob/2a2a76ad6540177af19cbb4194895496978d288a/packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts#L197-L206

Which was added in #1847 by @G-Rath but I'm not sure why.

**Comment by G-Rath** [comment]
@MatthiasKunnen see #7227

**Comment by mdwiltfong** [comment]
Is this still an issue? I see that the code is now throwing an error in the ts playground. 

cc: @JoshuaKGoldberg

**Comment by JoshuaKGoldberg** [comment]
Brad and I chatted briefly - we're slowly starting to sour on this rule as there are a couple issues (#6547, #7227) that are surprisingly intricate with it. Removing `good first issue`.

I also filed #8113 to track reworking the rule more holistically. I'm going to close & lock this issue for now so that we can focus on that discussion. If it goes nowhere I'll re-open this one. Thanks all!
